### PR TITLE
Fix project build for VS2015

### PR DESCRIPTION
**Pull request by omtinez**
This change addresses the required build configuration changes for this project to build in the latest version of Visual Studio. Summary of changes:
- Use platform-independent `os.path.join` and `os.sep` instead of hardcoded directory separators
- Add `settimeout` function to the list of automatically generated header files. This is necessary when trying to compile with some of the components under /lib
- Remove two digit exponent directive for Visual Studio versions 2015 and above, since now it's the default and it would cause a build failure.
- Update output location of binaries in the `vcxproj` file to a more standard location, rather than the project root folder

Tested and working for Win32 and x64.

### CODE REVIEW

**Comment by omtinez** [comment]
This fixes the issues described in https://github.com/micropython/micropython/issues/1532

File: py/makeversionhdr.py [hunk 1]
```diff
@@ -57,7 +57,7 @@ def get_version_info_from_git():
     return git_tag, git_hash, ver
 
 def get_version_info_from_docs_conf():
-    with open("%s/docs/conf.py" % sys.argv[0].rsplit("/", 2)[0]) as f:
+    with open(os.path.join(sys.argv[0].rsplit(os.sep, 2)[0], "docs", "conf.py")) as f:
```
Comment (stinos) [inline py/makeversionhdr.py]: This will still fail if a path with forward slashes is passed to the script. As @pfalcon mentioned earlier, the correct way is using os.path.dirname()

File: unix/qstrdefsport.h [hunk 2]
```diff
@@ -79,6 +79,7 @@ Q(listen)
 Q(accept)
 Q(recv)
 Q(recvfrom)
+Q(settimeout)
```
Comment (stinos) [inline unix/qstrdefsport.h]: If possible surround this with an ifdef checking for the feature this is needed for?

File: windows/micropython.vcxproj [hunk 3]
```diff
@@ -1,5 +1,5 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
-<Project DefaultTargets="Build" ToolsVersion="12.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+<Project DefaultTargets="Build" ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (stinos) [inline windows/micropython.vcxproj]: This, along with PlatformToolset, will make the project file unusable for earlier versions of VS (i.e. 2013). I knew the day would come we'd need to take care of dealing with muliple versions of VS (which is why the meat of the configuration is already in seperate .props files and the project is just a 'container' now) but never put much thought into how to do this exactly. Do you have any suggestions?

File: windows/micropython.vcxproj [hunk 4]
```diff
@@ -21,31 +21,32 @@
   <PropertyGroup Label="Globals">
     <ProjectGuid>{740F3C30-EB6C-4B59-9C50-AE4D5A4A9D12}</ProjectGuid>
     <RootNamespace>micropython</RootNamespace>
+    <WindowsTargetPlatformVersion>8.1</WindowsTargetPlatformVersion>
```
Comment (stinos) [inline windows/micropython.vcxproj]: Is this necessary?

File: windows/micropython.vcxproj [hunk 5]
```diff
@@ -75,10 +76,25 @@
   <PropertyGroup Label="UserMacros">
     <CustomPropsFile Condition="'$(CustomPropsFile)'==''">msvc/user.props</CustomPropsFile>
   </PropertyGroup>
-  <PropertyGroup />
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">
+    <OutDir>$(PyOutDir)$(Configuration)\$(Platform)\</OutDir>
+  </PropertyGroup>
```
Comment (stinos) [inline windows/micropython.vcxproj]: Compiler/linker configuration should go in the corresponding props file, common.props in this case, in order not to repeat yourself for all possible configuration/platform combos.
In any case it's debatable whether the output directory should change: all other ports use the same principle, and I guess there's a good reason for it @pfalcon @dpgeorge ?

File: windows/micropython.vcxproj [hunk 6]
```diff
@@ -75,10 +76,25 @@
...
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'">
+    <OutDir>$(PyOutDir)$(Configuration)\$(Platform)\</OutDir>
+  </PropertyGroup>
   <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">
     <ClCompile />
     <Link />
+    <ClCompile />
+    <ProjectReference>
+      <UseLibraryDependencyInputs>false</UseLibraryDependencyInputs>
+    </ProjectReference>
```
Comment (stinos) [inline windows/micropython.vcxproj]: This seems a leftover from something else?

**Comment by stinos** [comment]
Since you have 4 pretty much distinct changes, it is probably best to make 4 different pullrequests for it.

File: unix/qstrdefsport.h (hunk 2 above)
Comment (omtinez) [inline unix/qstrdefsport.h]: This isn't really a header file (in fact, I don't like that the extension is .h). It is being parsed programatically by py/makeqstrdata.py to generate a proper header file so no ifdef's are possible at this stage. The other header file that references the generated header file is also being created programatically, so it's a catch-22

File: windows/micropython.vcxproj (hunk 4 above)
Comment (omtinez) [inline windows/micropython.vcxproj]: It is not, and why VS added it on my machine running Windows 10 is a mystery to me.

File: windows/micropython.vcxproj (hunk 6 above)
Comment (omtinez) [inline windows/micropython.vcxproj]: Ah, yes. False is the default. I was playing around with this trying to get the linking to work.

File: windows/micropython.vcxproj (hunk 5 above)
Comment (omtinez) [inline windows/micropython.vcxproj]: Yep, I didn't notice it was already defined in the props file. I am unfamiliar with the other ports but dumping the binaries in the same folder as the rest of the code sounds like a pretty bad idea to me: hard to do cleanup and confusing. I thought that building had failed because I couldn't find the binaries folder!

File: windows/micropython.vcxproj (hunk 3 above)
Comment (omtinez) [inline windows/micropython.vcxproj]: The usual methodology is to keep this up to date with the latest version and the project members that still have an older one need to change this file by hand prior to opening the project and exclude their project files from their commits. It's unfortunate that VS does not handle this scenario better, specially considering that newer versions can typically open an older solution and the changes occur automatically.

**Comment by omtinez** [comment]
Considering that 1) is a relatively small change (< 15 lines), 2) it all goes towards the same feature (enabling building project for Visual Studio 2015), and 3) for everyone's sanity; I would strongly prefer to keep this as a single pull request.

File: unix/qstrdefsport.h (hunk 2 above)
Comment (dpgeorge) [inline unix/qstrdefsport.h]: Not true.  All qstrdef*.h files go through the C preprocessor first, then through makeqstrdata.py.  So you can use #if's.  See line 153 of py/py.mk.  (Maybe this is not true though for windows build?)

Comment (pfalcon) [inline unix/qstrdefsport.h]: This is definitely true for the official Window build system (GNU Makefile). This is apparently true for MSVC2013 build system, or how it can build stuff at all?

But the main question, how is that possible that MSVC2013 (and any other compiler) doesn't need "settimeout" qstr, and MSVC2015 needs it?

Comment (omtinez) [inline unix/qstrdefsport.h]: This was only the case when compiling with the lwIP module, after I removed that dependency I never tested it without this line.

Comment (stinos) [inline unix/qstrdefsport.h]: >  This is apparently true for MSVC2013 build system, or how it can build stuff at all?

indeed the header is ran through the preprocessor in ```genhdr.targets```

File: windows/micropython.vcxproj (hunk 3 above)
Comment (stinos) [inline windows/micropython.vcxproj]: Would it not make more sense to just support the oldest version then, since upgrading can be done automatically and effortless?

File: py/makeversionhdr.py [hunk 7]
```diff
@@ -57,7 +57,7 @@ def get_version_info_from_git():
     return git_tag, git_hash, ver
 
 def get_version_info_from_docs_conf():
-    with open("%s/docs/conf.py" % sys.argv[0].rsplit("/", 2)[0]) as f:
+    with open(os.path.join(os.path.dirname(sys.argv[0]), "docs", "conf.py")) as f:
```
Comment (stinos) [inline py/makeversionhdr.py]: argv[0] is some/path/to/micropython/py/makeversionhdr.py so you need to call dirname twice to make it to some/path/to/micropython

File: windows/micropython.vcxproj [hunk 8]
```diff
@@ -75,10 +75,10 @@
   <PropertyGroup Label="UserMacros">
     <CustomPropsFile Condition="'$(CustomPropsFile)'==''">msvc/user.props</CustomPropsFile>
   </PropertyGroup>
-  <PropertyGroup />
   <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">
     <ClCompile />
     <Link />
+    <ClCompile />
```
Comment (stinos) [inline windows/micropython.vcxproj]: Can you get rid of this line as well?

File: windows/micropython.vcxproj (hunk 3 above)
Comment (stinos) [inline windows/micropython.vcxproj]: > This, along with PlatformToolset, will make the project file unusable for earlier versions of VS (i.e. 2013).

Nevermind, seems I was wrong all along: opening the upgraded project in VS2013 just uses the v120 platformtoolset automatically so that worked out flawlessly after all..

Comment (omtinez) [inline windows/micropython.vcxproj]: So you can just double click the project file and it opens in Visual Studio 2013 just fine? That's news to me. Did it prompt you that the project could not be loaded? Did it prompt you to convert the project to a compatible VS version? Did the project file change to v120 everywhere after opening it?

File: py/makeversionhdr.py (hunk 7 above)
Comment (omtinez) [inline py/makeversionhdr.py]: You are correct, somehow I thought this worked since the build succeeded. How come `Build -> Clean` did not get rid of the generated files?

File: windows/micropython.vcxproj (hunk 3 above)
Comment (stinos) [inline windows/micropython.vcxproj]: > So you can just double click the project file and it opens in Visual Studio 2013 just fine?

Only if VS2015 is not installed, else you have to explicitely open in VS2013, but in any case it works ok on the 2 machines I tried it on. No prompt for downgrading, no changes to the project file. It's news to me as well, in fact I'm sure tons of people out there have no idea, but it's good news.

File: py/makeversionhdr.py (hunk 7 above)
Comment (stinos) [inline py/makeversionhdr.py]: Likely the build succeeded because ```get_version_info_from_docs_conf``` is only called if building 'bare' uPy sources and not when building git sources. The latter is the case probably since you're creating PRs..
Cleaning does not delete generated headers because I never bothered implementing that; I don't think it is needed currently: makeqstrdata.py and makeversionhdr.py are called for every build.

File: windows/micropython.vcxproj (hunk 3 above)
Comment (pfalcon) [inline windows/micropython.vcxproj]: Shouldn't it be the other way around - newer version of IDE should be able to open project files from older version without problems?

Comment (omtinez) [inline windows/micropython.vcxproj]: That's not the case, and other IDE's present the same issue (Android Studio comes to mind). Newer version of the IDE tries to convert the project to the latest version, if possible in a backwards-compatible way. Depending on the project type and version diff, converting the project might fail or it may end up being incompatible with older versions of the IDE.

Comment (pfalcon) [inline windows/micropython.vcxproj]: > That's not the case, and other IDE's present the same issue

How is that problem of the individual projects, MicroPython in this case? That's bug in the IDE, it should be reported to its vendor, and resolved on its side.

Comment (omtinez) [inline windows/micropython.vcxproj]: Project templates/settings change, not always in a compatible way. It's not a bug, it's a breaking change. In this case, all settings being exposed in the vcxproj of this project appear to be compatible between versions and the transition is seamless. Not all project types are the same, and the breaking changes are specially common in mobile application project types or unstable/beta types such as pyproj.

File: unix/qstrdefsport.h (hunk 2 above)
Comment (omtinez) [inline unix/qstrdefsport.h]: So the pre-processor does multiple passes? First, on this header file and then on the generated header file? 

What's the harm of adding this line? The way I see it, it allows to compile using the lwIP lib and many of the other lines in this file are (correct me if I'm wrong) not needed; I would assume that all unnecessary generated header files will be optimized away regardless.

Comment (dhylands) [inline unix/qstrdefsport.h]: Adding a Q(setblocking) line causes memory to be allocated to hold the qstr. So, the general idea is that for features which can be turned on or off then there should be an #ifdef around the qstr definitions as well, something like: https://github.com/micropython/micropython/blob/master/stmhal/qstrdefsport.h#L215-L245

Comment (omtinez) [inline unix/qstrdefsport.h]: Then what should be done about this? As far as I can tell, I can remove a decent portion of the lines in this file and building still succeeds. I have no idea of what the implications of that are on the final binaries. So adding an ifdef for a single line seems silly to me when some major cleanup is needed here.

Comment (pfalcon) [inline unix/qstrdefsport.h]: > Then what should be done about this?

This page should answer many questions:
https://github.com/micropython/micropython/wiki/ContributorGuidelines (link to it is shown on creation of pull request).

Comment (omtinez) [inline unix/qstrdefsport.h]: From that link, under the unix port: "every byte is still accounted for". That doesn't seem to apply here, since as I described I can remove quite a few lines and build still succeeds. The more important question would be *how that affects the final binary in terms of functionality*. If a single line is so important, then this file needs to be revised to remove all other unnecessary lines.

Comment (dhylands) [inline unix/qstrdefsport.h]: > Then what should be done about this? As far as I can tell, I can remove a decent portion of the lines in this file and building still succeeds. I have no idea of what the implications of that are on the final binaries. So adding an ifdef for a single line seems silly to me when some major cleanup is needed here.

After merging all of the qstrs from all of the places that generate qstrs, duplicates will be removed. So every unique Q(xxx) will be instantiated exactly once in the final binary.

So if FEATURE A needs Q(recv) and FEATURE B needs Q(recv) we should define it for both features. Otherwise, if FEATURE A is enabled and you didn't provide it with FEATURE B, and FEATURE A gets turned off, then FEATURE B will start to fail in the compile.

I'm sure that we're currently not doing this religiously, but that's the philosophy.

Comment (omtinez) [inline unix/qstrdefsport.h]: @dhylands: That makes a lot of sense. Is there a way to programmatically tell which feature needs which Q(xxx), provided that the (feature) header files that reference the header files generated from Q(xxx) are also generated programmatically as far as I can tell?

In a more concrete example, to be able to compile the Windows port using the lwIP lib I had to go down a wild goose chase for the header that was generated by X script and referenced a header generated by Y script; not a fan of code auto-generation but I guess it works. Eventually, I sort of *guessed* that the Q(settimeout) would give me the header that I needed. Is there a better approach than going down that path and adding an #ifdef for every possible lib after the problem is encountered?

Comment (stinos) [inline unix/qstrdefsport.h]: > So the pre-processor does multiple passes? First, on this header file and then on the generated header file? 

well yes, but the generated header file does not contain any ifdefs anymore, only the qstr definitions

> I have no idea of what the implications of that are on the final binaries

just that it will be smaller in size, afaik

> Then what should be done about this?

for instance unix/qstrdefs.h can be improved by surrounding all socket related qstrs with ```#if MICROPY_PY_SOCKET```, the ffi related qstrs with ```#if MICROPY_PY_FFI``` and so on

Comment (dhylands) [inline unix/qstrdefsport.h]: @omtinez So for headers in general - it's just the usual game as far as C headers go. 

For qstrs, each unique string gets assigned a unique integer which is then used in the byte code. In C code, you'll see something like this:
https://github.com/micropython/micropython/blob/master/stmhal/uart.c#L726-L751

Each of the MP_QSTR_xxx is from an auto-generated enum. So MP_QSTR_deinit would be generated in response to Q(deinit).

From my build tree, genhdr/qstrdefs.generated.h contains the line:
```
QDEF(MP_QSTR_deinit, (const byte*)"\x9e\x8d\x06" "deinit")
```

The genhdr/qstrdefs.generated.h is included once here (to create the enumeration):
https://github.com/micropython/micropython/blob/master/py/qstr.h#L39-L44
and once here (to allocate the string and hash):
https://github.com/micropython/micropython/blob/master/py/qstr.c#L90-L100

And there are a variety of sources of the qstrs. This file covers the qstrs from the py tree:
https://github.com/micropython/micropython/blob/master/py/qstrdefs.h
and then each port provides the port specific ones:
https://github.com/micropython/micropython/blob/master/stmhal/qstrdefsport.h

I don't see any module ones right now, but I imagine that there could be module specific files ones as well.

Comment (omtinez) [inline unix/qstrdefsport.h]: @dhylands: thanks for the explanation, I think I'm beginning to understand the magic behind the qstr's. Please correct me if I'm wrong, but as far as I can tell there is one "master" /py/qstrdefs.h and then, for each port, there *could* be a specific qstr definition that gets added in addition to the *master* qstr, since they just get processed in series as far as I can see in the props file: `<Exec Command="python $(PySrcDir)makeqstrdata.py $(PreProc) $(QstrDefs) > $(TmpFile)"/>` and the `do_work` function in `/py/makeqstrdata.py`. It doesn't seem, however, that there is a way to be able to concatenate more than two of those qstr headers (at least for the Windows project), so adding lib-specific qstr's would require hand-crafting the port-specific one.

In any case, it seems to me now like `Q(settimeout)`, being something needed for lwIP and not the Windows/unix port, **definitely** does not belong to the `/unix/qstrdefsport.h`. Would that be correct?

Comment (dhylands) [inline unix/qstrdefsport.h]: We don't have anything in place to deal with qstrs needed by an external module like lwIP.

For example, extmod/modubinascii needs some qstr's like hexlify, which currently comes from here: https://github.com/micropython/micropython/blob/master/py/qstrdefs.h#L597-L603

I took a look at the stmhal Makefile, and there are some other places we get qstr's from:
https://github.com/micropython/micropython/blob/master/stmhal/Makefile#L16

When we process the pins file for a particular board we generate all of the qstr's needed to define the pins, and similarly, for the stmmodule we actually parse some C header files to extract offsets and base addresses and we also generate qstr's that are needed for that.

I think that in an ideal world, that lwIP should have a qstrdefs.h file that describes all of the qstrs that it needs, and that the port that wants lwIp should then add the appropriate qstrdefs.h file to the list of files to be processed.

Comment (dhylands) [inline unix/qstrdefsport.h]: @omtinez And I see that there are already LWIP qstr's here:
https://github.com/micropython/micropython/blob/master/py/qstrdefs.h#L618-L645

and settimeout is already defined in there.

Comment (dpgeorge) [inline unix/qstrdefsport.h]: Indeed, current practice is to put extmod qstrs in py/qstrdefs.h.

**Comment by omtinez** [comment]
The commits have been squashed, since all comments have been addressed.

File: py/makeversionhdr.py [hunk 9]
```diff
@@ -1,4 +1,4 @@
-"""
+﻿"""
```
Comment (pfalcon) [inline py/makeversionhdr.py]: Please figure out what's wrong here and fix it.

**Comment by pfalcon** [comment]
Thanks, this looks much better, but as was pointed out earlier, each change which does something separate should go in a separate commit (for example, I'd be happy to apply 2 of the changes above right away, but the rest will need explicit confirmation of @stinos). I even looked a way to do git cherry-pick -p (which shamingly doesn't exist), but there's none which allows to preserve the original author.

File: py/makeversionhdr.py (hunk 9 above)
Comment (omtinez) [inline py/makeversionhdr.py]: Fixed

**Comment by omtinez** [comment]
@pfalcon I fail to see which changes in this PR are not part of something separate. All these changes enable building in Visual Studio. Are you suggesting one PR per file? Per line?

**Comment by pfalcon** [comment]
> All these changes enable building in Visual Studio.

Yes, and all commits in this repo can be squashed to one with commit message "MicroPython". And yet we do it differently, and there're even formal guidelines how to do it (so, it's not about your patch - it's about every patch).

> Are you suggesting one PR per file? Per line?

Per separate change, as guidelines explain. For example, I'd to 3 changes here: one to makeversionhdr.py
 to let it work properly with backslash paths (and commit message should say *that*, not some generic "in the name of"), another one to call _set_output_format() - note that it has nothing to do with *building*, so the commit message is plain wrong (I also could ask if you really confirmed it doesn't work for mingw, but I'm skipping that, as I'd really like to merge it). These 2 commits I'd be happy to merge right away, because they are pretty clear even for me, who doesn't run MSVC. The rest will need to wait for ack from @stinos as the MSVC build system maintainer.

Sorry if it seems so hard - it doesn't have to be, and guidelines were written to avoid obvious problems. So I'm wondering just the same, why it's so hard to follow pretty common-sense rule - don't throw your veggies mixed up into a fridge, keep each type separate in its own box, for easy access and maintenance.

**Comment by omtinez** [comment]
> call _set_output_format() - note that it has nothing to do with building, so the commit message is plain wrong

Actually, due to MSVC 14.0 breaking change (as explained in the link which is part of the comments above the line change) building *does* fail without that #ifdef which, by the logic implied, should not affect mingw.

I understand that there are rules and guidelines for a reason, but I also believe that there is an incremental improvement between project build failing and project build working. I also presumed that the windows port, being labeled as *experimental*, would not be subject of such a strict (almost dogmatic?) review process when none of the changes affect other ports. 

The bottom line is that I wanted a binary for different Windows flavors; the Windows project was completely broken and I found a way to fix it for myself which required 1) the backslash fix 2) the output format flag and 3) the binary output folder structure. I thought that I would give back to the project by sharing what the fix was, adjusting for the feedback, and adhering to the project guidelines to the best of my abilities.

Being pretty new to Git, and with the problem I was encountering solved for me, I'm afraid that I don't have the time to open 3 separate pull requests and, frankly, that seems quite unreasonable for a ~10 line change. 

> and commit message should say that, not some generic "in the name of"

That is good feedback, more coherent and verbose commit messages is something I should be more diligent about.

**Comment by dpgeorge** [comment]
@omtinez it would help if you could separate the changes into 2 or 3 separate commits, for each logical thing (esp separating out the change to py/makeversionhdr.py).  Then we can merge them individually and you don't need 3 separate PRs.

Of course we could just make the changes ourselves, but it's nice to have your name as the author of the commit in the history of the git log (and I hope that's something you want!).

**Comment by omtinez** [comment]
@dpgeorge I think that I was able to break it out into the discrete commits, let me know if my branch now contains what you requested.

@pfalcon do you think that the discrete commits adhere to the project guidelines better or would you still prefer independent pull requests? Any comments/thoughts about the commit messages?

@stinos what do you think about the changes in the MSVC project? As stated in the commit message, having a different binary output path for different project configurations is the only way to build for different flavors without overwriting the output. Additionally, the produced binaries do not get dumped in the same folder as code and it is better aligned with the default project configuration in Visual Studio.

**Comment by pfalcon** [comment]
> @pfalcon do you think that the discrete commits adhere to the project guidelines better or would you still prefer independent pull requests? Any comments/thoughts about the commit messages?

No independent pull requests are needed (in this case). I've merged https://github.com/micropython/micropython/commit/17c649da3d406f72907f46acc9473bbeebb30d91 and https://github.com/micropython/micropython/commit/3510499a9e523e51b1d8015f7d393aea9b68f679 , indeed adjusted commit messages a bit, hope you're ok with that. Thanks for your patches and patience! You may want to rebase your branch and re-push it (with --force switch), so only remaining patches were shown here.

@stinos : Please ack that the rest is ok to go. As a minor nitpick, there's:

~~~~
+    <OutDir>$(PyOutDir)$(Configuration)\$(Platform)\</OutDir>
     <IntDir>$(PyBuildDir)$(Configuration)$(Platform)\</IntDir>
~~~~

(see backslash difference). I'm personally ok with it, just want to be sure it doesn't break anything ;-)

**Comment by stinos** [comment]
> being labeled as experimental

good point, maybe it's time to remove that label, after all the port works fine, passes almost all tests and imo has enough features to be used as a small/fast python implementation. Also the README explicitely mentions VS2013 so after these changes get merged that should change as well

> such strict (almost dogmatic?) review process when none of the changes affect other ports

experimental or not, being strict (as in, basically deciding for each single line changed whether it is ok for the project in the longer run) is probably the only way to maintain a public open source project without turning it into a mess. Also other ports not being affected is orthogonal to other users/developers not being affected so that point is moot.

> do you think that the discrete commits adhere to the project guidelines better

much better, but still some minor things left: commit messages should be prefixed with the directory (I never used this before contributing to uPy but it grew onto me now and it actually makes a lot of sense) and should, if possible, also explain *why* a change was made, next to what was changed (the former documents the latter so PR reviewers know what the commit is about, the latter does not always need to be stated again since the diff does that already), so your first message could for example be something like

 ```
py: Replace forward slashes with OS independent os.path calls

This fixes errors in makeversionhdr.py when using paths paths
with backward slashes on Windows.
```

Also note that github deals with such format nicely when creating a PR: first line is the title of thr PR, other lines go in the PR description.

Regarding the individual changes:
- _set_output_format change is ok, as stated in the VS2015 changelog you linked to
- project update is ok, as it enables build using VS2015 and appears to not affect VS2013, and older versions lack the C99 support to build uPy anyway so they don't matter

Anyway, just saw that pfalcon merged the above already - including proposed changes - while I was writing this; I've some things to take care of now and I'll be back later with comments on the output dir change

**Comment by pfalcon** [comment]
Few comments from me too:

> building does fail without that #ifdef which, by the logic implied, should not affect mingw.

That's why review process - you see, my implied logic doesn't match your implied logic, if you said it clearly (preferrably in commit message), there wouldn't be more questions.

> The bottom line is that I wanted a binary for different Windows flavors; the Windows project was completely broken

It was not! We build Windows as part of our Continuous Integartion loop, and the badge on https://github.com/micropython/micropython in README is always (well, 99.9%) green.

> none of the changes affect other ports.

makeversionhdr.py change does affect.

> for a ~10 line change.

It's a small change because that stuff was already polished pretty much, the more we're interested to understand why more changes are still required. Because otherwise it's easy to go over top of the hill and actually deteriorate it.

> I'm afraid that I don't have the time to open 3 separate pull requests

We don't do all this to for bureaucracy's sake - we hate it very much (speaking for myself, I contribute  patches to various projects, and sometimes less than satisfied with the process, so trying my best to do it better for uPy). So, no 3 pull requests are required (3 commits - yes). But would you be able in 6 months to answer why each of the changes was made in the original patch you submitted with the original description? Sorry, I doubt that (and 6 months is optimistic figure, most real-world multitasking people won't be able in 1 month). Now imagine that in 5 month comes some guy, saying "project completely broken" and pointing at your commit. You may imagine a possible outcome - the complete commit may get reverted, because indeed, it's questionable. With our process, questions resolved first, and whoever may come later saying that it's completely broken, will get patient hints that it's not, and more homework is required to actually isolate issue(s) and find teh best way to fix them without affecting what's already worked for people so far (and otherwise reshuffling stuff). Hope all that is actually something to appreciate.

**Comment by dpgeorge** [comment]
Thanks @omtinez for splitting it into 3 commits.

For your interest, and others following the discussion, here is a random selection of commits with detailed messages, sometimes the message is bigger than the commit: bedab235f9028562b6677fa257cd08fbf1efecca dfa915a6af714d6d6f185c0d5a67cd45a9f7acd1 d80174d7c3fe36f139c639d6a820cc4cbb88ddf9 01d64914c5517f4207c002d931efa1e71fc210d9

The project is at a stage now where lots of people use and rely on it and changing some code can have an unforeseen impact (and changing code related to the build process is one area that can easily cause problems).  To mitigate such problems we need to document the development well, and that comes through proper git commit messages.

**Comment by stinos** [comment]
> Please ack that the rest is ok to go.

- project change: ok for me, though ideally it would be accompanied by a modification of the port's README so it states the new functionality
- output directory change: not ok in it's current state after some investigation
 - it's different from what all (?) the other ports use
 - as such it also breaks running tests with default settings (1)
 - not a fan of $(Configuration)\$(Platform) format, there are enough directories already, without the slash is better imo
 - build is easy enough to customise for that or any other output directory the user wants (2)
 - my mistake, sorry for that, but the change should be made in env.props not common.props - reason being the first is meant to actually declare all directories and is included by multiple other files

(1) Now you just run ```python run-tests``` no matter what platform/configuration used and no matter whether building on unix, windows mingw or windows msvc. To accomodate this change however, either one has to use e.g. ```set MICROPY_MICROPYTHON=..\windows\Debug\Win32\micropython.exe``` first (and hope that matches the platform/config last built) or the default path used in run-tests has to be changed (but then we'd have to fight over what that default should be.. and in any case that would break test running for mingw builds). None of these are nice solutions, especially given they are solutions to a problem which didn't exist in the first place. @omtinez out of interest: which tests fail using the VS2015 builds?

(2) Place user.props in msvc directory and adjust any build setting to taste, it's ignored by git btw

However after checking the current situation I'm also not too happy with the windows dir being littered with map/pdb/whatever files and I agree it might be handy to have different exes simultaneously. Also the build output now goes to windows/msvc/build while mingw build output goes to windows/build (which is consistent with other ports). 

Long story short: I'd fix this using a 'best of both worlds approach' putting msvc executables in windows/build as well, using $(Configuration)$(Platform) subdirectory, and object files in again a subdirectory of that, and add a post-build step which copies the executable to the windows directory no matter which flavor is built. This seems the only way to address as much as possible of the considerations raised by all parties. Please comment. @omtinez if this suits you but you don't have time let me know and I'll be on it.

**Comment by omtinez** [comment]
> You may want to rebase your branch and re-push it

Done.

> the README explicitely mentions VS2013 so after these changes get merged that should change as well

I agree with this, I did not notice that.

> my implied logic doesn't match your implied logic, if you said it clearly (preferrably in commit message), there wouldn't be more questions.

This was miscommunication on my part. What I meant by implied logic was the the `#elif` can't, by the logic of the `#ifdef`, affect mingw.

>> the Windows project was completely broken

> It was not! We build Windows as part of our Continuous Integartion loop

I am very surprised by this, considering the state of the build process when I first attempted it on VS2015.

> But would you be able in 6 months to answer why each of the changes was made in the original patch you submitted with the original description? Sorry, I doubt that

Fair enough, I never questioned that. The single-PR and single-commit is a fairly common way of managing external contributions. Realistically, not many times there is a need to have granularity to the individual commit level. Obviously, it's a matter of preference.

> To mitigate such problems we need to document the development well, and that comes through proper git commit messages.

See above. While I personally agree with more granular information, most other projects prefer single-commit and single-PR. That's my failure for not adhering to the stated project guidelines and making assumptions instead.

@stinos : the project configuration is more about standards than about personal preference. Adhering to the platform standards makes it much easier for new developers/users to understand it from the get-go. Maybe not as important as some might tell you, but certain tools depend on (hope might be a better word?) standard binary output directories. I, for one, have a number of scripts that call MSBUILD and fish the appropriate binaries out of the expected output folder.

I can move the changes from common.props to env.props, but before I do that I will take a look at the situation with the tests and report back.

Thank you all for your patience with this, I'm happy to be able to contribute to this project.

**Comment by stinos** [comment]
> standards 

There isn't really a standard here, just a default used by one IDE on one particular platform. *If* there's one standard way to figure out where any build system puts it's output files then the standard is looking at the build output or inspecting the build files. Users, especially developers, or scripts should know better than to expect things, they better be sure. As an anecdotical exercise I went through all projects which I currently have on my pc (I'll name some known ones for illustration: CPython, Boost, LLVM, CLang, NuGet) , and none of them uses the default.

**Comment by omtinez** [comment]
> Users, especially developers, or scripts should know better than to expect things, they better be sure

In theory. In practice, any project worth its salt inevitably evolves into having a somewhat complex build system and making assumptions about the outputs is the only sensible way to tackle some issues. I know of at least one project where post-build test running breaks after the binary output location is changed :-) What matters to me is that I can call MSBUILD with certain parameters and expect the output to be in a certain location which will not be overwritten by subsequent MSBUILD calls with different parameters. If it's in the default location, then it saves me the trouble of having to look at the actual project files to see what goes where.

> CPython, Boost, LLVM, CLang, NuGet

Those are all aimed to be platform-independent; I'd say that, in my anecdotal experience, I have rarely seen Windows-only projects or projects that started as Windows-only and don't follow the IDE default. Removing one level of nested folders like you suggest is OK with me but I personally don't find it as clean as the default. I guess my question should be more like *is there a good reason not to use the default?* In this case, it breaks uniformity with respect to the other ports, which qualifies as a solid reason in my book.

**Comment by stinos** [comment]
> What matters to me is that I can call MSBUILD with certain parameters and expect the output to be in a certain location which will not be overwritten by subsequent MSBUILD calls with different parameters. If it's in the default location, then it saves me the trouble of having to look at the actual project files to see what goes where.

Fair enough, the main thing I have against the default is that it puts one directory per configuration right in the source, that's almost on the same level as just putting all output files in the source dir. I care less whether it has extra nested subdirs so just see if putting whatever subdir format under the build directory suits you.

Btw the default VS uses here is $(SolutionDir)\$(Platform)\$(Configuration)\, so the last two are the reverse of your suggestion?

**Comment by stinos** [comment]
So as it happens just yesterday I ran into linker errors when switching between build configurations, because all output goes into one single directory. The problem has always been there of course but it went largely unnoticed and only now I happened to need a ton of memory which lead to switching back and forth between x86/x64 builds. @omtinez https://github.com/stinos/micropython/commit/fb8939a2ca1bd7a7f1411a7d764362695bc2804b is the proposed fix which is basically the idea I laid out in the above comments somewhere.

**Comment by omtinez** [comment]
@stinos I am looking over the commit that you linked, please tell me if I got it right:

1. The output directory is structured in the way that you proposed, e.g. windows\DebugWin32\*
2. After the build, the resulting .exe is copied into windows\ like with the other ports

This seems to me like the best compromise. It's not adding any more nested directories than strictly necessary, and the output is both where one would expect it to be (+/- the slightly different directory structure) and consistent with the other ports.

**Comment by stinos** [comment]
> e.g. windows\DebugWin32*

actually it's ```windows\build\DebugWin32```, so all build artifacts go under the build directory. That seemed cleaner to me than polluting the windows directory with one subdirectory per build type (now that's 'only' 4, but suppose uPy can be built into a static library by selecting configurations with a different name than it blows up) plus after all that's what the build directory is meant for.

**Comment by omtinez** [comment]
Sounds good to me. Was that the last issue to be addressed by this pull request? Should I close it now?

**Comment by stinos** [comment]
Not yet: your commit with the upgraded project file hasn't been merged into master yet. If you remove your change to common.props and rebase then @pfalcon can merge this. After which this PR and #1532 can be closed and VS2015 support is fixed.

**Comment by omtinez** [comment]
@stinos: Please take a look, I think that I was able to rebase with the latest changes and removed my edits to common.props

**Comment by stinos** [comment]
Looks good! Ok for me to merge this.

**Comment by pfalcon** [comment]
Merged, thanks guys!
